apps/s_client.c

```diff
@@ -798,6 +801,31 @@ static int new_session_cb(SSL *S, SSL_SESSION *sess)
     return 0;
 }
 
+static void keylog_callback(const SSL *ssl, const char *line)
+{
+    BIO *keylog = BIO_new_file(keylog_file, "a");
+    char buffer[256];
```

Comment:
My thinking with the newline at the time was that you might have multiple `SSL_CTX`s in the same application that write to the same file, maybe even concurrently. In that case, your callback is going to funnel into some code (or be that code) which merges entries from each of them.

Failing to terminate an entry with a newline would be a framing error in the serialization, so it seemed cleaner to say the merging code (which would normally be responsible for keeping the two sources from colliding by way of, say, threading) should be responsible for preventing framing errors. To that end, it seemed better for the merging code to be responsible for framing (i.e. \n separators) than for the merging code to assume things were safe to concatenate.

Does putting it in one `write` call actually work? Can't `write` return that it only managed to write part of the data and then you'd have this problem anyway?